File: OOD Java/Assignment8/src/main/java/polyhedra/PolyhedronFactory.java
package polyhedra;

/**
 * This is the Polyhedron Creating Wizard.
 */
public final class PolyhedronFactory {

    /**
     * List of known Polyhedra.
     */
    private static Polyhedron[] knownPolyhedra = {
        new Sphere(),
        new Cylinder(),
        new Composite()
    };

    /**
     *  Create a Polyhedron.
     *
     *  @param name the Polyhedron to be created
     *
     *  @return A Polyhedron with the specified name
     *      or null if no matching Polyhedron is found
     */
    public static Polyhedron createPolyhedron(final String name)
    {
        for (Polyhedron polyModel : knownPolyhedra) {
            // if (polyModel.getType().toLowerCase().equals(name)) {
            if (polyModel.getType().equalsIgnoreCase(name)) {
                return polyModel.clone();
            }
        }

        return null;
    }

    /**
     *  Determine whether a given Polyhedron is known.
     *
     *  @param name the Polyhedron for which to query
     *
     *  @return true if name is a known polyhedron
     */
    public static boolean isKnown(final String name)
    {
        for (Polyhedron polyModel : knownPolyhedra) {
            // if (polyModel.getType().toLowerCase().equals(name)) {
            if (polyModel.getType().equalsIgnoreCase(name)) {
                return true;
            }
        }

        return false;
    }

    /**
     *  Print a list of known Polyhedra.
     *
     *  @return string containing the names of all known polyhedra, one per
     *  line
     */
    public static String listKnown()
    {
        StringBuilder bld = new StringBuilder();

        for (Polyhedron polyModel : knownPolyhedra) {
            bld.append("  " + polyModel + "\n");
        }

        return bld.toString();
    }

    /**
     *  Determine the number of known Polyhedra.
     *
     *  @return the number of known Polyhedra
     */
    public static int numberKnown()
    {
        return knownPolyhedra.length;
    }

    //--------------------------------------------------------------------------
    /**
     * Do not allow PolyhedronFactory objects to be created.
     */
    private PolyhedronFactory()
    {
    }
}

File: OOD C++/Assignment4/PolyhedronFactory.cpp
#include "PolyhedronFactory.h"

#include "Sphere.h"
#include "Cylinder.h"

#include <iterator>
#include <algorithm>

PolyhedronFactory::PolyhedronPair PolyhedronFactory::_known_polyhedra[] = {
    {"sphere"   ,  new Sphere()},
    /*PolyhedronPair( "composite",  new Composite()), // This will be in part 2*/
    {"cylinder" ,  new Cylinder()}
};

//------------------------------------------------------------------------------
PolyhedronFactory::PolyhedronPair::PolyhedronPair()
    : _name(), _model(nullptr)
{
}

//------------------------------------------------------------------------------
PolyhedronFactory::PolyhedronPair::PolyhedronPair(std::string name, Polyhedron* polygon)
    : _name(name), _model(polygon)
{
}

PolyhedronFactory::PolyhedronPair::~PolyhedronPair()
{
    delete _model;
}

//------------------------------------------------------------------------------
Polyhedron* PolyhedronFactory::createPolyhedron(std::string name)
{
    for(const PolyhedronPair& pair : _known_polyhedra) {
        if(pair._name == name) {
            return pair._model->clone();
        }
    }

    // A polygon with the given name could not be found
    return nullptr;
}

//------------------------------------------------------------------------------
bool PolyhedronFactory::isKnown(std::string name)
{
    /*
    for(const PolyhedronPair& pair : _known_polyhedra){
        if(pair._name == name){
            return true;
        }
    }

    // The polygon with the given name is unknown
    return false;
    */

    const auto it = std::find_if(std::begin(_known_polyhedra),
                                 std::end(_known_polyhedra),
                                 [&name](const PolyhedronPair& pairToCheck) {
                                     return pairToCheck._name == name;
                                 });

    return it != std::end(_known_polyhedra);
}

//------------------------------------------------------------------------------
void PolyhedronFactory::listKnown(std::ostream& outs)
{
    for(const PolyhedronPair& pair : _known_polyhedra) {
        outs << " " << pair._name << "\n";
    }
}

//------------------------------------------------------------------------------
int PolyhedronFactory::numberKnown()
{
    int count = 0;

    for(const PolyhedronPair& pair : _known_polyhedra) {
        count++;
    }

    return count;
}
